src/mtree.erl

```diff
@@ -0,0 +1,515 @@
+    verify_uncle_hash(Tree, Hash_List, root_hash(Tree), {Bin, Hash}).
+
+verify_uncle_hash(Tree, Hash_List, {Root_Bin, Root_Hash}, {Bin, Hash})
+  when Bin rem 2 =:= 0 ->
+    case verify_uncle_hash(Tree, Hash_List, {Root_Bin, Root_Hash},
+                           {Bin, Hash}, []) of
+        %% Uncle_Hashes also includes all the intermediate parents calculated
+        %% using the actual uncle hashes
+        {ok, true, Uncle_Hashes} ->
+            {ok, insert_uncle_hashes(Tree, Uncle_Hashes)};
+        {ok, false, _} ->
+            {error, mtree_uncle_hash_mismatch};
+        {error, Msg, _} ->
+            {error, Msg}
+    end;
+verify_uncle_hash(_, _, _, _) ->
+    {error, mtree_not_a_leaf}.
+
+verify_uncle_hash(_Tree, [], {R_Bin, R_Hash}, {R_Bin, Hash}, Acc) ->
+    {ok, mtree_core:compare(R_Hash, Hash), Acc};
+verify_uncle_hash(Tree, [], {R_Bin, R_Hash}, {Bin, Hash}, Acc) ->
+    {Status, Value} = verify(Tree, {R_Bin, R_Hash}, {Bin, Hash}),
+    {Status, Value, Acc};
+verify_uncle_hash(Tree, Hash_List, {R_Bin, R_Hash}, {Bin, Hash}, Acc) ->
+    Sibling =  mtree_core:get_sibling(Bin),
+    case get_hash(Hash_List, Sibling) of
+        error        ->
+            {error, mtree_missing_uncle_hashes_in_list, empty};
+        Sibling_Hash ->
+            Parent_Hash   = mtree_core:hash([Hash, Sibling_Hash]),
+            Parent_Bin    = erlang:round((Bin + Sibling)/2),
+            New_Hash_List = lists:keydelete(Sibling, 1, Hash_List),
+            New_Acc = [{Sibling, Sibling_Hash},
+                       {Parent_Bin, Parent_Hash} | Acc],
+
+            verify_uncle_hash(Tree, New_Hash_List,{R_Bin, R_Hash},
+                              {Parent_Bin, Parent_Hash}, New_Acc)
+    end.
+
+%% HELPER fun
+%% get hash from the Hash_List
+get_hash(Hash_List, Bin) ->
+    case lists:keyfind(Bin, 1, Hash_List) of
+        {_, Hash} -> Hash;
+        false     -> error
+    end.
+
+%% HELPER fun
+%% insert Uncle_Hashes into the tree
+insert_uncle_hashes(Tree, Hash_List) ->
+    lists:map(fun({Bin, Hash}) ->
+                      mtree_store:insert(Tree,{Bin, Hash, empty})
+              end, Hash_List).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc verify/2 for a given bin bumber and its hash it recalculates the root
+%% hash And verifies its against the root hash
+%% @end
+%% TODO thorough checking.
+-spec verify(mtree(), bin_hash(), bin_hash()) -> {ok, true}
+                                               | {ok, false}
+                                               | {error, _}.
+verify(_, {Root_Bin, Root_Hash}, {Bin, Hash}) when Root_Bin =:= Bin ->
+    {ok, mtree_core:compare_hash(Root_Hash, Hash)};
+verify(Tree, {Root_Bin, Root_Hash}, {Bin, Hash}) ->
+    Sibling = mtree_core:get_sibling(Bin),
+    case mtree_store:lookup(Tree, Sibling) of
+        {ok, Sibling_Hash, _} ->
+            Parent_Bin  = erlang:round((Bin + Sibling)/2),
+            Parent_Hash = if
```

Comment:
Replace the 'if' expression on line 420 with a 'case' expression or function clauses.